tests/produce_integration_test_data.py

```diff
@@ -524,42 +532,27 @@ def print_failure_stats(test, truth, inputs, abs_tol, rel_tol, name):
         return False
 
     failed_idx = np.where(sel_failed)
-    warnings.warn(
-        f"{name}: atol {abs_tol} rtol {rel_tol} failed {sel_failed.sum()} of {sel_failed.size} {sel_failed.sum() / sel_failed.size * 100:.4f}%",
-        stacklevel=2,
-    )
-    warnings.warn(
-        f"subcube of failures [min] [max] {[f.min() for f in failed_idx]} {[f.max() for f in failed_idx]}",
-        stacklevel=2,
-    )
-    warnings.warn(
-        f"failure range truth ({truth[sel_failed].min():.3e},{truth[sel_failed].max():.3e}) test ({test[sel_failed].min():.3e},{test[sel_failed].max():.3e})",
-        stacklevel=2,
-    )
-    warnings.warn(
-        f"max abs diff of failures {np.fabs(truth - test)[sel_failed].max():.4e} relative {(np.fabs(truth - test) / truth)[sel_failed].max():.4e}",
-        stacklevel=2,
-    )
+    message = f"{name}: atol {abs_tol} rtol {rel_tol} failed {sel_failed.sum()} of {sel_failed.size} {sel_failed.sum() / sel_failed.size * 100:.4f} %\n"
+    message += f"subcube of failures [min] [max] {[f.min() for f in failed_idx]} {[f.max() for f in failed_idx]}\n"
+    message += f"failure range truth ({truth[sel_failed].min():.3e},{truth[sel_failed].max():.3e}) test ({test[sel_failed].min():.3e},{test[sel_failed].max():.3e})\n"
+    message += f"max abs diff of failures {np.fabs(truth - test)[sel_failed].max():.4e} relative {(np.fabs(truth - test) / truth)[sel_failed].max():.4e}\n"
```

Comment:
This actually might be easier to read if formatted as a multiline string, i.e.

```python
message = f"""
{name}: atol {abs_tol} ... etc
subcube of failures [min] [max] .... etc
"""
```

Reply:
I cleaned it up a bit using implicit concatenation, but the indentation on the multiline strings makes this a bit annoying.